packages/boot/src/controller.booter.ts

```diff
@@ -0,0 +1,129 @@
+// Copyright IBM Corp. 2013,2018. All Rights Reserved.
+// Node module: @loopback/boot
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {CoreBindings, Application, Booter, BootOptions} from '@loopback/core';
+import {inject} from '@loopback/context';
+import * as glob from 'glob';
+
+/**
+ * ControllerBooter is a class that implements the Booter inferface. The purpose
+ * of this booter is to allow for convention based booting of `Controller`
+ * artifacts for LoopBack 4 Applications.
+ *
+ * It supports the following boot phases: config, discover, boot
+ */
+export class ControllerBooter implements Booter {
+  options: ControllerOptions;
+  projectRoot: string;
+
+  /**
+   *
+   * @param app Application instance
+   * @param bootConfig Config options for boot
+   */
+  constructor(
+    @inject(CoreBindings.APPLICATION_INSTANCE) public app: Application,
+    @inject(CoreBindings.BOOT_CONFIG) public bootConfig: BootOptions,
+  ) {
+    if (!bootConfig.controllers) bootConfig.controllers = {};
+    this.options = bootConfig.controllers;
+    this.projectRoot = bootConfig.projectRoot;
+  }
+
+  /**
+   * This phase is responsible for configuring this Booter. It converts
+   * options and assigns default values for missing values so ther phases
+   * don't have to perform checks / conversions.
+   */
+  async config() {
+    this.options = Object.assign({}, ControllerDefaults, this.options);
+    if (typeof this.options.dirs === 'string') {
+      this.options.dirs = [this.options.dirs];
+    }
+    if (typeof this.options.extensions === 'string') {
+      this.options.extensions = [this.options.extensions];
+    }
+  }
+
+  /**
+   * This phase is responsible for discovering artifact files based on the
+   * given parameters. Sets options.discovered to an array of discovered
+   * artifact files.
+   */
+  async discover() {
+    const dirs = <string[]>this.options.dirs;
+    const exts = <string[]>this.options.extensions;
```

Comment:
These explicit type casts look ugly to me.

I am proposing to store the normalized options as first-level properties of the booter class.

```ts
class ControllerBooter {
  dirs: string[];
  extensions: strings[];

  async config() {
    const options = Object.assign({}, ControllerDefaults, this.options);
    this.dirs = normalizeToArray(options.dir);
    this.extensions = normalizeToArray(options.dir);
  }
}

function normalizeToArray<T>(scalarOrArray: T | T[]): T[] {
  return Array.isArray(scalarOrArray) ? scalarOrArray : [scalarOrArray];
}
``